core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,439 @@
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *BundlePool) filter(tx *types.Transaction) bool {
+	switch tx.Type() {
+	case types.LegacyTxType, types.AccessListTxType, types.DynamicFeeTxType:
+		return true
+	default:
+		return false
+	}
+}
+
+func (p *BundlePool) reset(oldHead, newHead *types.Header) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *BundlePool) addBundles(bundles []*types.Bundle) []error {
+	errs := make([]error, len(bundles))
+	for i, bundle := range bundles {
+		if err := p.AddBundle(bundle); err != nil {
+			errs[i] = err
+		}
+	}
+	return errs
+}
+
+func (p *BundlePool) drop() {
+	leastPrice := big.NewInt(math.MaxInt64)
+	leastBundleHash := common.Hash{}
+	for h, b := range p.bundles {
+		if b.Price.Cmp(leastPrice) < 0 {
+			leastPrice = b.Price
+			leastBundleHash = h
+		}
+	}
+	p.PruneBundle(leastBundleHash)
+}
+
+// =====================================================================================================================
+
+// NewBundleGasPricer creates a new BundleGasPricer.
+func NewBundleGasPricer(expire time.Duration) *BundleGasPricer {
+	return &BundleGasPricer{
+		expire: expire,
+		queue:  prque.New[int64, *gasPriceInfo](nil),
+		latest: common.Big0,
+	}
+}
+
+// BundleGasPricer is a limited number of queues.
+// In order to avoid too drastic gas price changes, the latest n gas prices are cached.
+// Allowed as long as the user's Gas Price matches this range.
+type BundleGasPricer struct {
+	mu     sync.RWMutex
+	expire time.Duration
+	queue  *prque.Prque[int64, *gasPriceInfo]
+	latest *big.Int
+}
+
+type gasPriceInfo struct {
+	val  *big.Int
+	time time.Time
+}
+
+// Push is a method to cache a new gas price.
+func (pool *BundleGasPricer) Push(gasPrice *big.Int) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	pool.retire()
+	index := -gasPrice.Int64()
+	pool.queue.Push(&gasPriceInfo{val: gasPrice, time: time.Now()}, index)
+	pool.latest = gasPrice
+}
+
+func (pool *BundleGasPricer) retire() {
+	now := time.Now()
+	for !pool.queue.Empty() {
+		v, _ := pool.queue.Peek()
+		info := v
+		if info.time.Add(pool.expire).After(now) {
+			break
+		}
+		pool.queue.Pop()
+	}
+}
+
+// LatestBundleGasPrice is a method to get the latest-cached bundle gas price.
+func (pool *BundleGasPricer) LatestBundleGasPrice() *big.Int {
+	pool.mu.RLock()
+	defer pool.mu.RUnlock()
+	return pool.latest
+}
+
+// MinimalBundleGasPrice is a method to get minimal cached bundle gas price.
+func (pool *BundleGasPricer) MinimalBundleGasPrice() *big.Int {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if pool.queue.Empty() {
+		return common.Big0
+	}
+	pool.retire()
+	v, _ := pool.queue.Peek()
+	return v.val
+}
+
+// Clear is a method to clear all caches.
+func (pool *BundleGasPricer) Clear() {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	pool.queue.Reset()
+}
+
+// numSlots calculates the number of slots needed for a single bundle.
+func numSlots(bundle *types.Bundle) uint64 {
```

Comment:
do we have reference code for this logic?

Reply:
The legacy pool shares same logic